### PR TITLE
Switch to Dugite for Cloning

### LINKED ISSUE #52: Switch to Dugite for cloning
Update: NodeGit is also a pain to build on Windows because of issues with node-gyp. Pushing priority of this up for MVP Release because it'll require more often to figure out how to get NodeGit to work instead of just switching to Dugite.

Nodegit is causing significant performance bottlenecks when cloning multiple repositories. Dugite is a package with native git bindings for Electron that was developed by the GitHub team.

### TODO:
- [x] Refactor `cloneutils.js` for Dugite
- [x] Parse system output and callback with progress
- [x] Measure performance against nodegit
- [x] Fix submission folder bug
- [x] Add tests for Progress Parser

### PR DESCRIPTION
**Pull request by srinjoym**
Fixes #52 

Switches to using Dugite (native Git bindings) for cloning. In addition to improved performance, this allows us to actually build the app on windows because issues with NodeGit compiling on Windows 🙌

### CODE REVIEW

File: app/lib/__tests__/cloneutils-test.js [hunk 1]
```diff
@@ -1,24 +1,25 @@
...
-import rmdir from "rimraf"
 
-const fs = require("fs")
+const fs = require("fs-extra")
 
 const TEST_REPO = "https://github.com/education/classroom-desktop"
 const TEST_FAKE_REPO = "https://github.com/education/a-repo-that-will-never-exist"
 
-const DESTINATION = "/tmp/" + Math.random().toString(36).substring(7)
+const DESTINATION = "/tmp/" + (Math.random().toString(36) + "00000").substr(2, 5)
```
Comment (damaneice) [inline app/lib/__tests__/cloneutils-test.js]: What do you think about extractiing `00000` into a CONST?

File: app/lib/cloneutils.js [hunk 2]
```diff
@@ -25,51 +25,25 @@ import _ from "underscore"
...
+  const result = await GitProcess.exec(
+    ["clone", repoURL, "--progress", "."],
+    destination,
+    options
+  )
+  // TODO Better Error handling and recovery
+  if (result.exitCode === 0) {
+    progressCallback(100)
+  } else {
+    throw new Error("Git Clone Process Failed.")
```
Comment (damaneice) [inline app/lib/cloneutils.js]: What happens when this fails? Is there an error message to the user?

File: app/lib/pathutils.js [hunk 3]
```diff
@@ -1,7 +1,18 @@
...
+  return path.join(basePath, `${assignmentName}-${time}`)
+}
+
+export const getClonePath = async (basePath, studentUsername) => {
+  const submissionPath = path.join(basePath, studentUsername)
+  try {
+    await fs.ensureDir(submissionPath)
+    return submissionPath
+  } catch (error) {
+    return null
```
Comment (damaneice) [inline app/lib/pathutils.js]: What happens when trap this error?   It looks like in the submissionViewDirectory function we try and open the path if it is not equal to an empty string. I think the path would be `null` in this scenario

```      const submission = candidates[0]
      if (submission.clonePath !== "") {
        opener(submission.clonePath)
      }

```

File: app/lib/cloneutils.js (hunk 2 above)
Comment (srinjoym) [reply app/lib/cloneutils.js]: The particular submission's status gets set to errored and we replace the progress bar with an error message!

File: app/lib/pathutils.js (hunk 3 above)
Comment (srinjoym) [reply app/lib/pathutils.js]: Just added a check for an error from this method in `submission-clone`, if this fails I display `Clone failed: Folder could not be created.` for that submission

**Comment by srinjoym** [comment]
@damaneice Updated this PR, is this good to go?

**Pull request merged**